File: v8-core/src/oracle/support.rs
//! Support classifier and evaluability validation (TARGET_ORACLE_SPEC §5.7,
//! §8, §16.2).
//!
//! Evaluates whether candidate/action pairs are supported by declared data,
//! execution, and environment models without fabricating point estimates.

#![allow(dead_code)]

use serde::{Deserialize, Serialize};

use crate::hash::Canon;
use crate::regret::Action;

use super::authority::CounterfactualAuthority;
use super::opportunity::GrammarCandidate;
use super::taxonomy::{AuthorityLevel, Identifiability, OracleRefusal};

#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
pub struct SupportRule {
    pub rule_id: String,
    pub environment_model_id: String,
    pub max_allowed_authority: AuthorityLevel,
    pub min_future_horizon_bars: usize,
    pub allowed_actions: Vec<String>,
    pub allows_model_counterfactuals: bool,
    pub assumptions: Vec<String>,
}

impl SupportRule {
    pub fn identity(&self) -> String {
        let mut c = Canon::new();
        c.push_str("support-rule-v1");
        c.push_value(&serde_json::json!({
            "rule_id": self.rule_id,
            "environment_model_id": self.environment_model_id,
            "max_allowed_authority": format!("{:?}", self.max_allowed_authority),
            "min_future_horizon_bars": self.min_future_horizon_bars,
            "allowed_actions": self.allowed_actions,
            "allows_model_counterfactuals": self.allows_model_counterfactuals,
            "assumptions": self.assumptions,
        }));
        c.finish_sha1_hex()
    }
}

pub struct SupportClassifier {
    pub rule: SupportRule,
}

impl SupportClassifier {
    pub fn new(rule: SupportRule) -> Self {
        Self { rule }
    }

    pub fn canonical_l1() -> Self {
        Self {
            rule: SupportRule {
                rule_id: "canonical-l1-support-v1".to_string(),
                environment_model_id: "binance-usdt-perp-l1".to_string(),
                max_allowed_authority: AuthorityLevel::L1,
                min_future_horizon_bars: 1,
                allowed_actions: vec![
                    "NO_TRADE".to_string(),
                    "ACTUAL".to_string(),
                    "GEOMETRY_VARIANT".to_string(),
                ],
                allows_model_counterfactuals: false,
                assumptions: vec![
                    "L1_BAR_CLOSE_FILL".to_string(),
                    "STATIC_SPREAD_FEE".to_string(),
                    "NO_ENDOGENOUS_IMPACT".to_string(),
                ],
            },
        }
    }

    /// Classify candidate and action support under requested authority level.
    pub fn classify(
        &self,
        candidate: &GrammarCandidate,
        action: &Action,
        requested_authority: AuthorityLevel,
    ) -> CounterfactualAuthority {
        let (auth, _) = self.evaluate_support(candidate, action, requested_authority, None);
        auth
    }

    /// Full evaluation returning orthogonal CounterfactualAuthority and optional OracleRefusal.
    pub fn evaluate_support(
        &self,
        candidate: &GrammarCandidate,
        action: &Action,
        requested_authority: AuthorityLevel,
        future_bars_available: Option<usize>,
    ) -> (CounterfactualAuthority, Option<OracleRefusal>) {
        // 1. Authority Level Check: L1 cannot support L2/L3 claims
        let authority_exceeded = matches!(
            (self.rule.max_allowed_authority, requested_authority),
            (AuthorityLevel::L1, AuthorityLevel::L2)
                | (AuthorityLevel::L1, AuthorityLevel::L3)
                | (AuthorityLevel::L1, AuthorityLevel::LiveReceipt)
                | (AuthorityLevel::L2, AuthorityLevel::L3)
                | (AuthorityLevel::L2, AuthorityLevel::LiveReceipt)
                | (AuthorityLevel::L3, AuthorityLevel::LiveReceipt)
        );

        if authority_exceeded {
            let mut assumptions = self.rule.assumptions.clone();
            assumptions.push("REQUESTED_AUTHORITY_EXCEEDS_SUPPORTED_MODEL".to_string());
            let auth = CounterfactualAuthority::new(
                requested_authority,
                Identifiability::NotIdentifiable,
                &self.rule.rule_id,
                &self.rule.environment_model_id,
                assumptions,
            );
            return (auth, Some(OracleRefusal::ExecutionAuthorityTooWeak));
        }

        // 2. Decision Time Data Check
        if candidate.decision_time < 0 {
            let auth = CounterfactualAuthority::new(
                requested_authority,
                Identifiability::NotIdentifiable,
                &self.rule.rule_id,
                &self.rule.environment_model_id,
                self.rule.assumptions.clone(),
            );
            return (auth, Some(OracleRefusal::MissingDecisionTimeData));
        }

        // 3. Action Kind Support
        if !self.rule.allowed_actions.iter().any(|a| a == action.kind || a == action.provenance) {
            let auth = CounterfactualAuthority::new(
                requested_authority,
                Identifiability::NotIdentifiable,
                &self.rule.rule_id,
                &self.rule.environment_model_id,
                self.rule.assumptions.clone(),
            );
            return (auth, Some(OracleRefusal::OutOfSupportAction));
        }

        // 4. Action Parameters & Fill Feasibility
        if let Some(size) = action.override_geom.get("size").and_then(|v| v.as_f64()) {
            if size <= 0.0 || !size.is_finite() {
                let auth = CounterfactualAuthority::new(
                    requested_authority,
                    Identifiability::NotIdentifiable,
                    &self.rule.rule_id,
                    &self.rule.environment_model_id,
                    self.rule.assumptions.clone(),
                );
                return (auth, Some(OracleRefusal::OutOfSupportAction));
            }
        }

        if action.override_geom.contains_key("queue_priority")
            || action.override_geom.contains_key("partial_fill_ratio")
        {
            // Queue and partial fill mechanisms require unobserved microstructure at L1
            let auth = CounterfactualAuthority::new(
                requested_authority,
                Identifiability::NotIdentifiable,
                &self.rule.rule_id,
                &self.rule.environment_model_id,
                self.rule.assumptions.clone(),
            );
            return (auth, Some(OracleRefusal::NonIdentifiableFill));
        }

        // 5. Future Horizon Check
        let expiry = action
            .override_geom
            .get("expiry_bars")
            .and_then(|v| v.as_u64())
            .map(|n| n as usize)
            .unwrap_or(self.rule.min_future_horizon_bars);

        if let Some(available) = future_bars_available {
            if available < expiry || available < self.rule.min_future_horizon_bars {
                let auth = CounterfactualAuthority::new(
                    requested_authority,
                    Identifiability::NotIdentifiable,
                    &self.rule.rule_id,
                    &self.rule.environment_model_id,
                    self.rule.assumptions.clone(),
                );
                return (auth, Some(OracleRefusal::UndefinedFuture));
            }
        }

        // 6. Model-only Counterfactual Check
        if action.override_geom.contains_key("model_counterfactual")
            && !self.rule.allows_model_counterfactuals
        {
            let auth = CounterfactualAuthority::new(
                requested_authority,
                Identifiability::ModelDerived,
                &self.rule.rule_id,
                &self.rule.environment_model_id,
                self.rule.assumptions.clone(),
            );
            return (auth, Some(OracleRefusal::ModelOnlyCounterfactual));
        }

        // 7. Fully Supported
        let auth = CounterfactualAuthority::new(
            requested_authority,
            Identifiability::Identified,
            &self.rule.rule_id,
            &self.rule.environment_model_id,
            self.rule.assumptions.clone(),
        );
        (auth, None)
    }
}

#[cfg(test)]
mod tests {
    use std::collections::BTreeMap;

    use super::*;
    use crate::oracle::opportunity::Direction;

    fn sample_candidate(decision_time: i64) -> GrammarCandidate {
        GrammarCandidate {
            grammar_candidate_id: "test-cand-01".to_string(),
            universe_id: "universe-v1".to_string(),
            template_id: "template-breakout".to_string(),
            instrument: "BTCUSDT".to_string(),
            timeframe: "1h".to_string(),
            direction: Direction::Long,
            decision_time,
            parameters: BTreeMap::new(),
        }
    }

    fn sample_action() -> Action {
        Action {
            action_id: "ACTUAL".to_string(),
            kind: "ACTUAL",
            provenance: "ACTUAL",
            override_geom: serde_json::Map::new(),
        }
    }

    #[test]
    fn l1_request_for_l3_returns_execution_authority_too_weak_and_not_identifiable() {
        let classifier = SupportClassifier::canonical_l1();
        let candidate = sample_candidate(100);
        let action = sample_action();

        let (auth, refusal) = classifier.evaluate_support(
            &candidate,
            &action,
            AuthorityLevel::L3,
            Some(10),
        );
        assert_eq!(auth.oracle_authority_level, AuthorityLevel::L3);
        assert_eq!(auth.identifiability_status, Identifiability::NotIdentifiable);
        assert_eq!(refusal, Some(OracleRefusal::ExecutionAuthorityTooWeak));
    }

    #[test]
    fn future_or_negative_decision_time_returns_missing_decision_time_data() {
        let classifier = SupportClassifier::canonical_l1();
        let candidate = sample_candidate(-1);
        let action = sample_action();

        let (auth, refusal) = classifier.evaluate_support(
            &candidate,
            &action,
            AuthorityLevel::L1,
            Some(10),
        );
        assert_eq!(auth.identifiability_status, Identifiability::NotIdentifiable);
        assert_eq!(refusal, Some(OracleRefusal::MissingDecisionTimeData));
    }

    #[test]
    fn unsupported_size_or_queue_fill_fails_support() {
        let classifier = SupportClassifier::canonical_l1();
        let candidate = sample_candidate(100);

        let mut bad_size_action = sample_action();
        bad_size_action
            .override_geom
            .insert("size".to_string(), serde_json::json!(-1.0));

        let (_, refusal1) = classifier.evaluate_support(
            &candidate,
            &bad_size_action,
            AuthorityLevel::L1,
            Some(10),
        );
        assert_eq!(refusal1, Some(OracleRefusal::OutOfSupportAction));

        let mut queue_action = sample_action();
        queue_action
            .override_geom
            .insert("queue_priority".to_string(), serde_json::json!(1));

        let (_, refusal2) = classifier.evaluate_support(
            &candidate,
            &queue_action,
            AuthorityLevel::L1,
            Some(10),
        );
        assert_eq!(refusal2, Some(OracleRefusal::NonIdentifiableFill));
    }

    #[test]
    fn insufficient_future_horizon_returns_undefined_future() {
        let classifier = SupportClassifier::canonical_l1();
        let candidate = sample_candidate(100);
        let mut action = sample_action();
        action
            .override_geom
            .insert("expiry_bars".to_string(), serde_json::json!(24));

        let (auth, refusal) = classifier.evaluate_support(
            &candidate,
            &action,
            AuthorityLevel::L1,
            Some(5), // only 5 bars available < 24
        );
        assert_eq!(auth.identifiability_status, Identifiability::NotIdentifiable);
        assert_eq!(refusal, Some(OracleRefusal::UndefinedFuture));
    }

    #[test]
    fn model_only_counterfactual_returns_model_only_refusal() {
        let classifier = SupportClassifier::canonical_l1();
        let candidate = sample_candidate(100);
        let mut action = sample_action();
        action
            .override_geom
            .insert("model_counterfactual".to_string(), serde_json::json!(true));

        let (auth, refusal) = classifier.evaluate_support(
            &candidate,
            &action,
            AuthorityLevel::L1,
            Some(10),
        );
        assert_eq!(auth.identifiability_status, Identifiability::ModelDerived);
        assert_eq!(refusal, Some(OracleRefusal::ModelOnlyCounterfactual));
    }
}

File: v8-next/src/v8_next/oracle/support.py
"""Support classifier and evaluability validation — port of v8-core/src/oracle/support.rs.

(TARGET_ORACLE_SPEC §5.7, §8, §16.2.)

Whether a candidate/action pair is supported by the declared data, execution, and
environment models — without fabricating point estimates. Unsupported profitable hindsight
paths are model-derived or unknown, never opportunities.

The ``Action`` shape mirrors ``regret::Action`` (the A(C) legal-action manifest); v8-next
carries no regret plane, so the minimal kind/provenance/geometry-overrides surface lives
here rather than as a parallel interface.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass, field
from typing import Any

from v8_next.oracle.authority import CounterfactualAuthority
from v8_next.oracle.opportunity import GrammarCandidate
from v8_next.oracle.taxonomy import AuthorityLevel, Identifiability, OracleRefusal

__all__ = [
    "Action",
    "SupportClassifier",
    "SupportRule",
]


@dataclass(frozen=True)
class Action:
    """One legal action from the A(C) manifest (kind/provenance/geometry overrides)."""

    action_id: str
    kind: str  # NO_TRADE | ACTUAL | GEOMETRY_VARIANT
    provenance: str  # ACTUAL | DECLARED_VARIANT
    override_geom: dict[str, Any] = field(default_factory=dict)


@dataclass(frozen=True)
class SupportRule:
    rule_id: str
    environment_model_id: str
    max_allowed_authority: AuthorityLevel
    min_future_horizon_bars: int
    allowed_actions: tuple[str, ...]
    allows_model_counterfactuals: bool
    assumptions: tuple[str, ...] = ()

    def identity(self) -> str:
        blob = json.dumps(
            {
                "rule_id": self.rule_id,
                "environment_model_id": self.environment_model_id,
                "max_allowed_authority": self.max_allowed_authority.name,
                "min_future_horizon_bars": self.min_future_horizon_bars,
                "allowed_actions": list(self.allowed_actions),
                "allows_model_counterfactuals": self.allows_model_counterfactuals,
                "assumptions": list(self.assumptions),
            },
            sort_keys=True,
            separators=(",", ":"),
        )
        return hashlib.sha256(f"support-rule-v1|{blob}".encode()).hexdigest()


def _refuse(
    rule: SupportRule,
    requested: AuthorityLevel,
    identifiability: Identifiability,
    refusal: OracleRefusal,
    extra_assumption: str | None = None,
) -> tuple[CounterfactualAuthority, OracleRefusal]:
    assumptions = list(rule.assumptions)
    if extra_assumption is not None:
        assumptions.append(extra_assumption)
    return (
        CounterfactualAuthority(
            oracle_authority_level=requested,
            identifiability_status=identifiability,
            support_rule_id=rule.rule_id,
            environment_model_id=rule.environment_model_id,
            assumptions=tuple(assumptions),
        ),
        refusal,
    )


class SupportClassifier:
    """Classify candidate/action support under a declared environment model."""

    def __init__(self, rule: SupportRule) -> None:
        self.rule = rule

    @classmethod
    def canonical_l1(cls) -> SupportClassifier:
        return cls(
            SupportRule(
                rule_id="canonical-l1-support-v1",
                environment_model_id="binance-usdt-perp-l1",
                max_allowed_authority=AuthorityLevel.L1,
                min_future_horizon_bars=1,
                allowed_actions=("NO_TRADE", "ACTUAL", "GEOMETRY_VARIANT"),
                allows_model_counterfactuals=False,
                assumptions=(
                    "L1_BAR_CLOSE_FILL",
                    "STATIC_SPREAD_FEE",
                    "NO_ENDOGENOUS_IMPACT",
                ),
            )
        )

    def classify(
        self,
        candidate: GrammarCandidate,
        action: Action,
        requested_authority: AuthorityLevel,
    ) -> CounterfactualAuthority:
        authority, _ = self.evaluate_support(candidate, action, requested_authority, None)
        return authority

    def evaluate_support(
        self,
        candidate: GrammarCandidate,
        action: Action,
        requested_authority: AuthorityLevel,
        future_bars_available: int | None,
    ) -> tuple[CounterfactualAuthority, OracleRefusal | None]:
        rule = self.rule
        # 1. Authority level check: a model cannot support claims above its level.
        exceeded = (
            (rule.max_allowed_authority is AuthorityLevel.L1 and requested_authority is not AuthorityLevel.L1)
            or (
                rule.max_allowed_authority is AuthorityLevel.L2
                and requested_authority
                in (AuthorityLevel.L3, AuthorityLevel.LIVE_RECEIPT)
            )
            or (
                rule.max_allowed_authority is AuthorityLevel.L3
                and requested_authority is AuthorityLevel.LIVE_RECEIPT
            )
        )
        if exceeded:
            return _refuse(
                rule,
                requested_authority,
                Identifiability.NOT_IDENTIFIABLE,
                OracleRefusal.EXECUTION_AUTHORITY_TOO_WEAK,
                "REQUESTED_AUTHORITY_EXCEEDS_SUPPORTED_MODEL",
            )
        # 2. Decision-time data check.
        if candidate.decision_time < 0:
            return _refuse(
                rule,
                requested_authority,
                Identifiability.NOT_IDENTIFIABLE,
                OracleRefusal.MISSING_DECISION_TIME_DATA,
            )
        # 3. Action-kind support.
        if not any(a == action.kind or a == action.provenance for a in rule.allowed_actions):
            return _refuse(
                rule,
                requested_authority,
                Identifiability.NOT_IDENTIFIABLE,
                OracleRefusal.OUT_OF_SUPPORT_ACTION,
            )
        # 4. Action parameters and fill feasibility.
        size = action.override_geom.get("size")
        if size is not None and (
            isinstance(size, bool) or not isinstance(size, (int, float)) or not size > 0.0
        ):
            return _refuse(
                rule,
                requested_authority,
                Identifiability.NOT_IDENTIFIABLE,
                OracleRefusal.OUT_OF_SUPPORT_ACTION,
            )
        if "queue_priority" in action.override_geom or "partial_fill_ratio" in action.override_geom:
            return _refuse(
                rule,
                requested_authority,
                Identifiability.NOT_IDENTIFIABLE,
                OracleRefusal.NON_IDENTIFIABLE_FILL,
            )
        # 5. Future-horizon check.
        expiry = action.override_geom.get("expiry_bars")
        if isinstance(expiry, bool) or not isinstance(expiry, int):
            expiry = rule.min_future_horizon_bars
        if future_bars_available is not None and (
            future_bars_available < expiry
            or future_bars_available < rule.min_future_horizon_bars
        ):
            return _refuse(
                rule,
                requested_authority,
                Identifiability.NOT_IDENTIFIABLE,
                OracleRefusal.UNDEFINED_FUTURE,
            )
        # 6. Model-only counterfactual check.
        if "model_counterfactual" in action.override_geom and not rule.allows_model_counterfactuals:
            return _refuse(
                rule,
                requested_authority,
                Identifiability.MODEL_DERIVED,
                OracleRefusal.MODEL_ONLY_COUNTERFACTUAL,
            )
        # 7. Fully supported.
        return (
            CounterfactualAuthority(
                oracle_authority_level=requested_authority,
                identifiability_status=Identifiability.IDENTIFIED,
                support_rule_id=rule.rule_id,
                environment_model_id=rule.environment_model_id,
                assumptions=rule.assumptions,
            ),
            None,
        )
